pkg/config/node.go

```diff
@@ -229,74 +219,40 @@ func GetNodes(kubeconfigPath string) (map[string][]v1.Node, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	for _, node := range nodes.Items {
-		// Get all labels for *each node* which contain node-role.kubernetes.io/
-		for label := range node.Labels {
-			switch {
-			case strings.HasPrefix(label, labelNodeRolePrefix):
-				if role := strings.TrimPrefix(label, labelNodeRolePrefix); len(role) > 0 {
-					nodeCluster[role] = append(nodeCluster[role], node)
-				}
-			}
-		}
-	}
-	return nodeCluster, nil
-}
-
-// IsTheSameConfig will compare the config annotations
-//
-// Args:
-//   - The node list based on v1.NodeList
-//
-// Returns:
-//   - true (same config) or false
-func IsTheSameConfig(nodes []v1.Node) bool {
-	desiredConfig := "machineconfiguration.openshift.io/desiredConfig"
-	currentConfig := "machineconfiguration.openshift.io/currentConfig"
-
-	for _, node := range nodes {
-		if node.Annotations[desiredConfig] != node.Annotations[currentConfig] ||
-			node.Annotations[desiredConfig] != nodes[0].Annotations[desiredConfig] {
-			return false
-		}
-	}
-	return true
+	return nodes, nil
 }
 
-// IsUpgradeStillRunning check if the upgrade still running in the nodes
-// comparing the node's state.
+// IsUpgradeStillRunning check if the upgrade is still running by looking at
+// the nodes' machineconfiguration state and kubelet version. Once all of the
+// machineconfigurations are Done and all kubelet versions match we know it
+// is safe to trigger the unicast migration.
 //
 // Args:
 //   - kubeconfigPath as string
 //
 // Returns:
-//   - true (different config - upgrade still running), false (upgrade complete) or error
+//   - true (upgrade still running), false (upgrade complete) or error
 func IsUpgradeStillRunning(kubeconfigPath string) (bool, error) {
 	nodes, err := GetNodes(kubeconfigPath)
 	if err != nil {
 		return false, err
 	}
 
 	kubeletVersion := ""
-	// Go to all node types identified in GetNodes()
-	for nodeRole := range nodes {
+	stateAnnotation := "machineconfiguration.openshift.io/state"
+	for _, node := range nodes.Items {
 		// Verify kubelet versions match. In EUS upgrades we may end up in an
 		// intermediate state where all of the nodes are "updated" as far as
 		// MCO is concerned, but are actually on different versions of OCP.
 		// In those cases, we do not consider the upgrade complete because not
 		// all nodes are ready for migration.
 		if kubeletVersion == "" {
-			kubeletVersion = nodes[nodeRole][0].Status.NodeInfo.KubeletVersion
+			kubeletVersion = node.Status.NodeInfo.KubeletVersion
 		}
-		if kubeletVersion != nodes[nodeRole][0].Status.NodeInfo.KubeletVersion {
+		if kubeletVersion != node.Status.NodeInfo.KubeletVersion {
```

Comment:
It may be not relevant at this moment, but I wonder if for heterogeneous clusters the `KubeletVersion` string is always the same. I think the answer is yes because I see `kubeletVersion: v1.25.4+77bec7a` so it contains version and git hash which would mean we have a platform for which kubelet was built based on a different commit, but nevertheless I want to raise this point because it may be useful in the future if we see something strange happening

Reply:
It's a good point and in a perfect world I would redesign all of this so we're not trying to guess what version all of the nodes are on, but at this point I don't think we can make drastic changes in 4.11. Of all the fields we have access to I think kubelet version is the least likely to vary within a cluster.

Reply:
I have read this now. Not sure if this comment will be of any help this late @cybertron , but just in case:
- In RHCOS, kubelet version is dictated by the RPM on the RHCOS version, so it is the same on all the RHCOS nodes.
- On the other hand, in RHEL nodes, it gets whatever kubelet version it can from the yum repos, so the kubelet version for RHEL nodes might be different.
- However, even if there is mismatch between RHEL nodes and RHCOS nodes, it would be only in the `.z` part, i.e. you may have a 1.24.a node and a 1.24.b node if one is RHCOS and the other one is RHEL, but you would never have a 1.24.z node and a 1.25.z node on a cluster unless it is in the middle of an upgrade (including an EUS upgrade) or somebody did something unsupported.